containers/jersey2-routing/src/main/java/com/yahoo/gondola/container/RoutingFilter.java

```diff
@@ -0,0 +1,234 @@
+import javax.ws.rs.container.ContainerRequestFilter;
+import javax.ws.rs.core.Response;
+
+public class RoutingFilter implements ContainerRequestFilter {
+
+    public static final String X_GONDOLA_LEADER_ADDRESS = "X-Gondola-Leader-Address";
+    public static final String APP_PORT = "appPort";
+    public static final String APP_SCHEME = "appScheme";
+    /**
+     * Routing table
+     * Key: memberId, value: HTTP URL
+     */
+    ClusterIdCallback clusterIdCallback;
+
+    Gondola gondola;
+    Set<String> myClusterIds;
+
+    // clusterId --> list of available servers
+    Map<String, List<String>> routingTable;
+
+    CloseableHttpClient httpclient;
+    public RoutingFilter(Gondola gondola, ClusterIdCallback clusterIdCallback) throws ServletException {
+        this.gondola = gondola;
+        this.clusterIdCallback = clusterIdCallback;
+        httpclient = HttpClients.createDefault();
+        loadRoutingTable();
+    }
+
+    @Override
+    public void filter(ContainerRequestContext requestContext) throws IOException {
+        String clusterId = getClusterId(requestContext);
+
+        if (isMyCluster(clusterId)) {
+            Member leader = getLeader(clusterId);
+
+            // Those are the condition that the server should process this request
+            // Still under leader election
+            if (leader == null) {
+                requestContext.abortWith(Response
+                                         .status(Response.Status.INTERNAL_SERVER_ERROR)
+                                         .entity("Under leader election")
+                                         .build());
+                return;
+            } else if (leader.isLocal()) {
+                return;
+            }
+        }
+
+        // Proxy the request to other server
+        proxyRequestToLeader(requestContext, clusterId);
+    }
+
+    private void loadRoutingTable() {
+        // The routing entry will be modified on the fly, concurrent map is needed
+        Map<String, List<String>> newRoutingTable = new ConcurrentHashMap<>();
+        Config config = gondola.getConfig();
+        for(String hostId : config.getHostIds()) {
+            if (hostId.equals(gondola.getHostId())) {
+                continue;
+            }
+            for(String clusterId : config.getClusterIds(hostId)) {
+                InetSocketAddress address = config.getAddressForHost(hostId);
+                List<String> addresses = newRoutingTable.get(clusterId);
+                if (addresses == null) {
+                    addresses = new ArrayList<>();
+                    newRoutingTable.put(clusterId, addresses);
+                }
+                Map<String, String> attrs = config.getAttributesForHost(hostId);
+                if (attrs.get(APP_PORT) == null || attrs.get(APP_SCHEME) == null) {
+                    throw new RuntimeException(
+                        String.format("Attribute '%s|%s' is required in gondola configuration. (gondola.hosts[].%s|%s)",
```

Comment:
"gondola.hosts[%s] is missing either the %s or %s config values", hostId, APP_PORT, APP_SCHEME.